### PR TITLE
Retina encoder: biological encoder for vision WIP

### PR DESCRIPTION
**Pull request by breznak**
eye.py is a biological implementation of retina (encoder).
WIP

For #682

### CODE REVIEW

**Comment by breznak** [comment]
FYI @ctrl-z-9000-times if you have some insights, please share, I'll slowly try to work this

File: py/htm/encoders/eye.py [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+
+    def reset(self):
+        self.retina.clearBuffers()
+
+
+class EyeSensorSampler:
+    """
+    Samples eyesensor.rgb, the eye's view.
+
+    Attribute samples is list of RGB numpy arrays.
```
Comment (breznak) [inline py/htm/encoders/eye.py]: make this accept a jpg/png image

File: py/htm/encoders/eye.py [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+        cols = math.ceil(num / rows)
+        for idx, img in enumerate(self.samples):
+            plt.subplot(rows, cols, idx+1)
+            plt.imshow(img[:,:,::-1], interpolation='nearest')
+        if show:
+            plt.show()
+
+
+# TODO: Consider splitting motor controls and motor sensory into different
+# classes...
```
Comment (breznak) [inline py/htm/encoders/eye.py]: :+1:

File: py/htm/encoders/eye.py [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+    max_change_angle        = (2*3.14159) / 500
+    eye_sensor.position     = (
+        eye_sensor.position[0] + random.gauss(1, .75),
+        eye_sensor.position[1] + random.gauss(1, .75),)
+    eye_sensor.orientation += random.uniform(-max_change_angle, max_change_angle)
+    eye_sensor.scale = 1
+
+
+if __name__ == '__main__':
+    eye = Eye()
```
Comment (breznak) [inline py/htm/encoders/eye.py]: make this a unit test, and an example (with visualizations)

File: py/htm/encoders/eye.py (hunk 2 above)
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: This code was part of a failed experiment and should be removed from this PR.  I do not fully understand how the brain controls the eye.

Instead the eye-encoder can just accept the coordinates, scale, and rotation of where the eye should look.

Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: I removed the motor control code.

File: py/htm/encoders/eye.py (hunk 1 above)
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: This should accept anything the python imaging library `PIL` accepts.

File: py/htm/encoders/eye.py [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+
+        self.retina = cv2.bioinspired.Retina_create(
+            inputSize            = (self.retina_diameter, self.retina_diameter),
+            colorMode            = True,
+            colorSamplingMethod  = cv2.bioinspired.RETINA_COLOR_BAYER,)
+
+        print(self.retina.printSetup())
+        print()
+
+        self.parvo_enc = encoders.ChannelEncoder(
```
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: Done, I copy-pasted it from old personal research repo into this file.

**Comment by ctrl-z-9000-times** [comment]
Thank you Breznak for getting the ball rolling on this PR.

I did some basic cleanup.  You should now be able to run this locally (without installing my old research repo).

The encoder is now run-able from the command line.  It requires a single argument: the image file-path or directory.
`$ python3 py/htm/encoders/eye.py ~/Pictures/`

File: py/htm/encoders/eye.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Thanh-Binh) [reply py/htm/encoders/eye.py]: where can I find your ChannelEncoder?

[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: The channel encoder is on line 25 of this file.

[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: And I wrote at length about how the "ChannelEncoder" works on the Numenta forum:
https://discourse.numenta.org/t/encoding-vision-for-htm/2780/18

File: py/htm/encoders/eye.py [hunk 5]
```diff
@@ -271,318 +410,69 @@ def input_space_sample_points(self, npoints):
...
+    else:
+        eye = Eye()
+        for img_path in images:
+            eye.reset()
+            print("Loading image %s"%img_path)
+            eye.new_image(img_path)
+            eye.scale = 1
+            for i in range(10):
+                pos,rot,sc = eye.small_random_movement()
+                sdr = eye.compute(pos,rot,sc) #TODO derive from Encoder
```
Comment (breznak) [inline py/htm/encoders/eye.py]: Q: @ctrl-z-9000-times how should the encode look? 
- `encode(image, pos, rot, scale)`
- `encode(image, n_steps=10, saccadic_move_fn=small_random_moves)`

With related problem below, the output SDR is very dense. Would logical AND over the SDRs from saccadic steps describe the overall image?

File: py/htm/encoders/eye.py [hunk 6]
```diff
@@ -30,58 +122,81 @@ class Eye:
...
-                            num_samples = 1, sparsity = sparsity ** (1/3.),
+        if mode == "both" or mode == "parvo":
+          dims = (output_diameter, output_diameter)
+          if color is True: 
+            dims = (output_diameter, output_diameter, 3,)
+
+          self.parvo_enc = ChannelEncoder(
+                            input_shape = dims,
+                            num_samples = 1, 
+                            sparsity = sparsity * (1/3.), #biologically, parvocellular pathway is only 33% of the magnocellular (in terms of cells)
```
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: this could be related to my problem with too high density: https://github.com/htm-community/htm.core/pull/691#discussion_r330530952

**Comment by breznak** [comment]
> validate the encode on MNIST classification task

I'm running into an import problem, `from htm.encoders.eye import Eye`, which is strange as GridCellEncoder works fine.

File: py/htm/encoders/eye.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: The reason the parvo-cellular has a weird sparsity is that there are three color channels, each of which is encoded separately and then combined.  The color channels are combined with a logical AND, which on average reduces the sparsity.

This same principal applies to any time you combine two encoders with a logical bit-wise AND.

File: py/htm/encoders/eye.py [hunk 7]
```diff
@@ -30,58 +122,81 @@ class Eye:
...
-    Attribute scale        ... 
+    Attribute orientation  ... units are radians TODO of what? sensor/image
+    Attribute scale        ... TODO
     """
     def __init__(self,
-        output_diameter   = 200,
-        resolution_factor = 3,
-        fovea_scale       = .177,
-        sparsity          = .2,):
+        output_diameter   = 200, #TODO set as percentage of imput image?
```
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: The output-diameter controls how many bits are in output encoded SDR, and I think that the output SDR should have a known size, irrespective of input image size.

Also, I intended for this encoder to be used a bit differently: the AI is supposed to control where in the image the eye is looking, so that the eye does not need to cover the whole image, it can search through the image.

File: py/htm/encoders/eye.py [hunk 8]
```diff
@@ -30,58 +122,81 @@ class Eye:
...
+    Attribute parvo ... SDR with parvocellular pathway (color)
+    Attribute magno ... SDR with magnocellular pathway (movement) 
 
     The following three attributes control where the eye is looking within
     the image.  They are Read/Writable.
     Attribute position     (X, Y) coords of eye center within image
-    Attribute orientation  ... units are radians
-    Attribute scale        ... 
+    Attribute orientation  ... units are radians TODO of what? sensor/image
+    Attribute scale        ... TODO
```
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: The scale controls the distance between the eye and them image.
I'm not sure what `scale`'s units are, or how to control it.  In my experiments I typically manually tuned the scale for the dataset, to find a reasonable value through trial and error.

File: py/htm/encoders/eye.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: > The reason the parvo-cellular has a weird sparsity is that there are three color channels, each of which is encoded separately and then combined.

good note, I'll add this as explanation for the 3rd root. Still, if we want the P-cells to have 1/3 sparsity of the M-cells, this should be `(sparsity/3) ** (1/3.)`

File: py/htm/encoders/eye.py [hunk 9]
```diff
@@ -30,58 +122,81 @@ class Eye:
...
-    Attribute magno ... 
+    Attribute parvo ... SDR with parvocellular pathway (color)
+    Attribute magno ... SDR with magnocellular pathway (movement) 
 
     The following three attributes control where the eye is looking within
     the image.  They are Read/Writable.
     Attribute position     (X, Y) coords of eye center within image
-    Attribute orientation  ... units are radians
-    Attribute scale        ... 
+    Attribute orientation  ... units are radians TODO of what? sensor/image
```
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: The eye can rotate, to look at the image sideways or upsidedown.  Actually, in a real animal, the *head* rotates, not the eyeball.

File: py/htm/encoders/eye.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: PS, I was thinking about replacing the `mode` with a `p_m_ratio` [0, 1.0]. We know the biological split is 1:3, but might be useful to have a continuous scale for experimantion? 
Then "parvo" would be p_m_ratio = 1, "magno" = p_m_ratio = 0, "both" (biological) would be 0.33, etc..

[the remark this answers is not in the document]
Comment (Thanh-Binh) [reply py/htm/encoders/eye.py]: @ctrl-z-9000-times: can we use only Magno?

File: py/htm/encoders/eye.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: >  In my experiments I typically manually tuned the scale for the dataset, to find a reasonable value through trial and error. 

hmm, so opencv's Retina model performance depends on the scale set.

File: py/htm/encoders/eye.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: > and I think that the output SDR should have a known size, irrespective of input image size.

agree, I was actually thinking to use "standard" output SDR size, and use the diameter only internally. 

> the AI is supposed to control where in the image the eye is looking, so that the eye does not need to cover the whole image, 

sure, that's the point. Might have a member `pct_area_covered` by "one look" computed from diameter and img size. But you're right, biologically that does not make sense, as the scene is continuous, and we look (around) as needed.

File: py/htm/encoders/eye.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: > can we use only Magno? 

depends on your task. 
see 
https://foundationsofvision.stanford.edu/chapter-5-the-retinal-representation/ 

> What type behaviors depend on the low spatial frequency and high temporal frequency information represented by the magnocellular pathway? Two examples of visual tasks that require precise and rapid information about rapidly varying image signals are motion detection and motion-tracking. The central projections of the magnocellular pathway we will review later, coupled with the significance of the perceptual signals, suggest that the magnocellular pathway plays an important role in providing high quality information used in motion perception.

TL;DR what I derive: 
```
magno ~ motion tracking, video
parvo   ~ image classification,... 
```

So if you have only a simple "objects" and want to detect movement, than magnocellular pathway would be used mainly.

[the remark this answers is not in the document]
Comment (Thanh-Binh) [reply py/htm/encoders/eye.py]: thanks

File: py/htm/encoders/eye.py [hunk 10]
```diff
@@ -1,27 +1,119 @@
...
+import cv2 # pip install opencv-contrib-python
+from PIL import Image # pip Pillow
+
+from htm.bindings.sdr import SDR
+
+
+class ChannelEncoder:
+    """
+    This assigns a random range to each bit of the output SDR.  Each bit becomes
+    active if its corresponding input falls in its range.  By using random
```
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: I wrote at length about this on the Numenta Forum.  I'm going to copy paste it all here, for us to edit and eventually move into the documentation.

Two requirements for encoding SDRs are that every bit of information represents a range of possible values, and that for every input multiple bits activate in the SDR. The effects of these requirements are that each output bit is inaccurate and redundant. This encoder makes every output bit receptive to a unique range of inputs, which are uniformly distributed throughout the input space. This meets the requirements for being an SDR and has more representational power than if many of the bits represented the identical input ranges. This design makes all of those redundancies add useful information.

Semantic similarity happens when two inputs which are similar have similar SDR representations. This encoder design does two things to cause semantic similarity: (1) SDR bits are responsive to a range of input values, and (2) topology allows near by bits to represent similar things.

1) Many encoders apply thresholds to real valued input data to convert the input into Boolean outputs. In this encoder uses two thresholds to form ranges which are referred to as ‘bins’. A small change in the input value might cause some of the output bits to change and a large change in input value will cause all of the output bits to change.  How sensitive the output bits are to changes in the input value -the semantic similarity- is determined by the sizes of the bins. The sizes of the bins are in turn determined by the sparsity, as follows:
    Assume that the inputs are distributed in a uniform random way throughout the input range. The size of the bins then determines the probability that an input value will fall inside of a bin. This means that the sparsity is related to the size of the bins, which in turn means that the sparsity is related to the amount of semantic similarity. This may seem counter-intuitive but this same property holds true for all encoders which use bins to convert real numbers into discrete bits.

2) This encoder relies on topology in the input image, the idea that adjacent pixels in the input image are likely to show the same thing.  If an area of the output SDR can not represent a color because it did not generate the bins needed to, then it may still be near to a pixel which does represent the color. In the case where there are less than one active output bits per input pixel, multiple close together outputs can work together to represent the input.  Also, if an output bit changes in response to a small change in input, then some semantic similarity is lost. Topology allows nearby outputs to represent the same thing, and since each output uses random bins they will not all change when the input reaches a single threshold.

Note about sparsity and color channels:
To encode color images create separate encoders for each color channel. Then recombine the output SDRs into a single monolithic SDR by multiplying them together. Multiplication is equivalent to logical “and” in this situation. Notice that the combined SDR’s sparsity is the different; the fraction of bits which are active in the combined SDR is the product of the fraction of the bits which are active in all input SDRs. For example, to make an encoder with 8 bits per pixel and a sparsity of 1/8: create three encoders with 8 bits per pixel and a sparsity of 1/2.

File: py/htm/encoders/eye.py (hunk 5 above)
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: > encode(image, pos, rot, scale)

Including the eye's positioning in the arguments is a good idea.  One of the reasons I stopped working with vision is because I don't know how to make the HTM control where in the image the eye is looking,  So it makes sense for the user to control the eye positioning.

File: py/htm/encoders/eye.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: Thanks, David :+1: 
I'll study this later on,

> This encoder makes every output bit receptive to a unique range of inputs, which are uniformly distributed throughout the input space. 

so a uniformly random subsample (of fixed size?) of bits is selected as receptive field for each output bit (?) 

- Numenta used to have a "RandomSP" mode, which was a SP without learning, only the initial random weights were used. Wouldn't that suffice the described behavior? 

> (2) topology allows near by bits to represent similar things. 

I haven't noticed any topology used in ChEnc? 

> This encoder relies on topology in the input image, the idea that adjacent pixels in the input image are likely to show the same thing.  

ok.. 

Overall: 

- this does seem like a useful way for a (generic) encoder
- [ ] can we use RDSE instead for this PR, and deal with `ChannelEncoder` separately? 
  - I somehow thought it's designed to deal with a stream of images, but what you describe sounds like an approach to a generic encoder

File: py/htm/encoders/eye.py (hunk 5 above)
Comment (breznak) [reply py/htm/encoders/eye.py]: > So it makes sense for the user to control the eye positioning. 

The question is the "level of management" we want to provide: 
- micro: `pos,rot,scale` represent positions in each of saccadic steps (ie. I'm looking at your left eye -> right eye -> mouth -> nose, ...) 
- macro: ditto represent "high level" position in the scene (ie I'm looking at "you"). 
  - and the encoder does (approximately) the saccadic movements (that is similar with our vision, we voluntarily control where we look at, but the micromanagement is subcontious). 

Intuitively, I'd be for the "macro" management mode, but how do we combine those sub-SDRs from saccades into one? logical-AND or OR?

File: py/htm/encoders/eye.py [hunk 11]
```diff
@@ -0,0 +1,479 @@
...
+            eye.reset()
+            print("Loading image %s"%img_path)
+            eye.new_image(img_path)
+            eye.scale = 1
+            for i in range(10):
+                pos,rot,sc = eye.small_random_movement()
+                sdr = eye.compute(pos,rot,sc) #TODO derive from Encoder
+                eye.show_view()
+                print("Sparsity: {}".format(len(sdr.sparse)/np.product(sdr.dimensions)))
+                print(sdr.dimensions) #TODO make SDR 2D, diameter x diameter. not current (200, 200, 2)
```
Comment (breznak) [inline py/htm/encoders/eye.py]: TODO make SDR 2D, diameter x diameter. not current (200, 200, 2). 

the 3rd dimensions represents P,M-cells? In that case, we should either: 
- drop the `mode` argument
- or keep mode and merge the SDR into a single layer (2D) 
  - I don't know if biologically the P,M-pathways are merged together, or lead to different regions?

File: py/htm/encoders/eye.py (hunk 5 above)
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: I think that this PR should do micro-level control over the eye movements.
1) Its what is already implemented
2) I don't really understand all of the science behind macro-level control

File: py/htm/encoders/eye.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: It's more about the photography of the images in the dataset.  When you look at a really big image, you need to take a few steps back to be able to see the whole thing.  The scale parameter allows the eye to do just that, walk forwards and backwards from the image, to find good place to view it from.

File: py/htm/encoders/eye.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: > pct_area_covered

The `scale` parameter should be used to control this value.

File: py/htm/encoders/eye.py (hunk 11 above)
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: > But does not conform.

That sounds like a bug!

---

Yes, the 3rd dimension is for P&M cells.

> I don't know if biologically the P,M-pathways are merged together, or lead to different regions?

IIRC: They project to different things in the midbrain & thalamus.  I'm not sure about what happens in the cortex.

File: py/htm/encoders/eye.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: Regarding the bits in output SDR, the ROI images (with blacked out corners, active inner circle), is the ROI only for visualization, or also the SDR is effectively not a square, but an inner circle?

File: py/htm/encoders/eye.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: >  Still, if we want the P-cells to have 1/3 sparsity of the M-cells, this should be (sparsity/3) ** (1/3.)

I cannot find where I read what the ratio is... 

> PS, I was thinking about replacing the mode with a p_m_ratio

and I was suggesting to replace the mode with p_m_ratio. I'd like to back that off, as the ratio would change the expected sparsity, which is given to constructor. 
Instead, we probably could have separate sparsities for `sparsityP`, `sparsityM`, and their (default) proportion would reflect the biological ratio of P,M-pathway sizes. 
Thoughts?

File: py/htm/encoders/eye.py (hunk 5 above)
Comment (breznak) [reply py/htm/encoders/eye.py]: `compute()` now expects coordinates for saccadic movements. 
Functionality is there, cannot derive from Encoder because #704

File: py/htm/encoders/eye.py (hunk 11 above)
Comment (breznak) [reply py/htm/encoders/eye.py]: > > But does not conform.

> That sounds like a bug!

Resolved, my bad, sorry! I computed the sparsity from both (the 3rd dimension) SDRs, now the computation is fixed and it gives 20% as expected. 

> Yes, the 3rd dimension is for P&M cells.

For clarity I added members 
```
self.parvo_sdr
self.magno_sdr
self.parvo_img
self.magno_img
```

File: py/htm/encoders/eye.py [hunk 12]
```diff
@@ -0,0 +1,529 @@
...
+        cv2.imshow('Region Of Interest', roi)
+        if self.color:
+          cv2.imshow('Parvocellular', self.parvo_img[:,:,::-1])
+        else:
+          cv2.imshow('Parvocellular', self.parvo_img)
+        cv2.imshow('Magnocellular', self.magno_img)
+        idx = self.parvo_sdr.dense.astype(np.uint8).reshape(self.output_diameter, self.output_diameter)*255
+        cv2.imshow('Parvo SDR', idx)
+        idx = self.magno_sdr.dense.astype(np.uint8).reshape(self.output_diameter, self.output_diameter)*255
+        cv2.imshow('Magno SDR', idx)
```
Comment (breznak) [inline py/htm/encoders/eye.py]: added a simple visualization for the magno/parvo SDRs

**Comment by breznak** [comment]
CC @fcr this could be interesting to you

**Comment by Thanh-Binh** [comment]
David: if I understood your ChannelEncoder exactly, it select an image pixel as active bit if and only if it’s intensity is in a given range [low, high].
If it is correct, it can not work with black-white images.
Am I right here?

**Comment by fcr** [comment]
> CC @fcr this could be interesting to you

Thanks [@breznak](https://github.com/breznak). I'll have a look.

**Comment by ctrl-z-9000-times** [comment]
> David: if I understood your ChannelEncoder exactly, it select an image pixel as active bit if and only if it’s intensity is in a given range [low, high].
> If it is correct, it can not work with black-white images.

Yes and No.

This encoder is designed to work with input pixels which have a continuous range of intensities.  If you know that all of your input data is boolean Black or White, then this is probably not a good choice of encoder.

However, it should still work, because there are bits which do respond to the colors black & white.  It might not work very well because many of the bits respond to the colors in between Black & White, ie greyscale, and they will never be activated by a black & white image.

File: py/htm/encoders/eye.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: The Region Of Interest (ROI, pretty pictures) is showing what the eye is looking at. The ROI images have not been encoded.

The encoding process involves doing a log-polar transform on the ROI, which discards the blacked out areas and outputs a rectangular shaped image.  The log-polar transform is the reason why the outputs of the encoder (Parvo & magno cellular) are warped.

Technically, the blacked out areas are approximately where the log-polar transform cuts off, but I do not know an exact formula for the radius of the blacked out areas.

**Comment by Thanh-Binh** [comment]
But it will work if instead of working on pixel intensity, we work on all pixels within a circle around each pixel. That means you select a pixel and it’s radius randomly. It is considered as active if the mean intensity for example of all pixels within this radius is over a threshold (defined by sparsity).
What do you think?

**Comment by ctrl-z-9000-times** [comment]
> What do you think?

I think that would be much more time consuming to compute.

File: py/htm/encoders/eye.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: FYI: it is P:M:other = 8:1:1 (atleast for makak monkey, which I take is good enough :) )

**Comment by breznak** [comment]
I think this is ready for general use, please review. 
The tests would need more love (resolution, colors, ... for Parvo; motion,speed sensitivity for Magno), but I'll leave that for other PRs.

File: py/tests/encoders/eye_test.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+""" Unit tests for retina encoder. """
+
+import unittest
+import os
+import numpy as np
+
+from htm.encoders.eye import Eye
+
+class EyeEncoderTest(unittest.TestCase):
+  """ Unit tests for Eye encoder class. """
```
Comment (ctrl-z-9000-times) [inline py/tests/encoders/eye_test.py]: These unit tests have no assert statements ......

File: py/htm/encoders/eye.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: > can we use RDSE instead for this PR, and deal with ChannelEncoder separately? 

No, because the channel encoder can encode each input scalar (pixel) into as few as 1 bit, but I think that the RDSE would break if told to encode an input scalar into 1 bit.

File: py/htm/encoders/eye.py [hunk 14]
```diff
@@ -0,0 +1,596 @@
...
+            centers = np.random.uniform(min(pad_drange),
+                                        max(pad_drange),
+                                        size=self.output_shape)
+        # Make the lower and upper bounds of the ranges.
+        low  = np.clip(centers - radius, min(self.drange), max(self.drange))
+        high = np.clip(centers + radius, min(self.drange), max(self.drange))
+        self.low  = np.array(low, dtype=self.dtype)
+        self.high = np.array(high, dtype=self.dtype)
+
+    def encode(self, img):
```
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: done for output, had some troubles with input as SDR, so I'm leaving that for later. Anyway, Channel is tightly coupled with Retina/image data - expects local similarity, log-pol encoding, ... 
so I think it's fine to expect an image as input here (?) 

Actually, I was thinking it'd be good if the log-pol transform was a part of this encoder (so Eye can be purely bioinspired.Retina) ...?

File: py/tests/encoders/eye_test.py (hunk 13 above)
Comment (breznak) [reply py/tests/encoders/eye_test.py]: yeah, this is really just a "does not crash"/API compatible test. 

I saw really nice examples for lightning conditions (parvo) and moving objects (magno), but that would be for later.

**Comment by breznak** [comment]
> Overall this looks good. I think it's close to good enough for this PR.  

thank you for reviewing! Addressed some of your concerns, I have more cleanups in another PR, but I broke sth there..so it's probably a good idea to get this into a shippable shape and resolve, and then do followups.

**Comment by breznak** [comment]
TODOs: 
- [ ] I don't know why the test on Linux fails in the CI https://github.com/htm-community/htm.core/pull/691/checks?check_run_id=270681996#step:11:169 (not really descriptive), passes locally. 
- [ ] resolve logpol without cropping to a ROI? #691 
- [ ] import non-breaking improvements from #691 
- [ ] test on MNIST
- [ ] complex unit tests for features of Retina (lighting, motion).

**Comment by breznak** [comment]
logpolar transform vs. retinal log sampling: 
https://github.com/opencv/opencv_contrib/issues/2305 

- both achieve "highlight in fovea, background less important"
- logpolar has the property that rotation,scale are represented as translations, thus have similar representations in encoder (overlap), but what about translation? 
- retinal log sampling is biologically plausible by the model. 
- both suffer that we (have to?) crop the image to ROI, and apply the transform there (with fovea in center of image/ROI), this cuts off the info outside of ROI, which would have been reduced anyway but continuously 

Tl;DR: can we use Retina.useRetinalLogSampling instead of manual logpolar transform?

File: py/htm/encoders/eye.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: > Actually, I was thinking it'd be good if the log-pol transform was a part of this encoder (so Eye can be purely bioinspired.Retina) ...?

I disagree. The retina does should do the log-polar transform, and the channel encoder should only do the transform from `pixels -> SDR`.  The Channel encoder could be used with other types of sensory data.

**Comment by ctrl-z-9000-times** [comment]
> logpolar has the property that rotation,scale are represented as translations, thus have similar representations in encoder (overlap), but what about translation?

Yes. Assuming the motion is small, the eye's output should still have semantic similarity.

> both suffer that we (have to?) crop the image to ROI, and apply the transform there (with fovea in center of image/ROI), this cuts off the info outside of ROI, which would have been reduced anyway but continuously.

No.  The area outside of the ROI is lost, not reduced.  The things outside of the ROI are outside of the eye's field of view.  The peripheral vision needs to be included inside of the ROI.

File: py/htm/encoders/eye.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: > The Channel encoder could be used with other types of sensory data.

are there any requirements on the data (similarity, encoding) for Channel to be used? Ie can I use it as a replacement for RDSE for encoding floats?

**Comment by breznak** [comment]
> The things outside of the ROI are outside of the eye's field of view. The peripheral vision needs to be included inside of the ROI. 

ok, there might be some micommunication of terms on my side, but imagine this case: 
"right now I'm reading on the notebook screen with the room in the background":
  - this is my field of view (FOV, about 160deg horizontally), that could be expressed as a photo taken from my position
  - my focus (fovea) is on the screen I read text from, which is only a tiny fration of the area on the picture
  - Q:  ROI==FOV vs.  ROI=="fovea"(screen)? 

This should illustrate that even for ROI (as implemented, the area of image that gets processed, other gets lost), if: 
  - ROI is FOV: we need to be able to specify position, diemeter of retinal fovea. 
  - ROI is "fovea", I think this is a bad design,as it leads to either: too much high-details (whole room in fovea), or almost none peripheral vision (only the "screen" and corners that fit into its bounding box), in reality I am able to notice motion in a large area. 

I think to summarize, 

> The peripheral vision needs to be included inside of the ROI. 

if this is true, we need to be able to specify the ratio of fovea/peripheral better

File: py/htm/encoders/eye.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ctrl-z-9000-times) [reply py/htm/encoders/eye.py]: This encoder needs to know the data range, so it can not replace the RDSE.  It could in theory replace the scalar encoder but I would not recommend it, since the scalar encoder works great and has nice deterministic properties.

[the remark this answers is not in the document]
Comment (breznak) [reply py/htm/encoders/eye.py]: ok, I'll make the split to standalone file + test (maybe as a part of this PR?)

**Comment by ctrl-z-9000-times** [comment]
The ROI is the _entire_ field of view.
The fovea is a small area at the center of the ROI.

> we need to be able to specify the ratio of fovea/peripheral better

This is one of the many tuning parameters, IIRC `self.fovea_scale`

File: py/htm/examples/mnist.py [hunk 15]
```diff
@@ -115,7 +120,10 @@ def main(parameters=default_parameters, argv=None, verbose=True):
     # Training Loop
     for i in range(len(train_images)):
         img, lbl = random.choice(training_data)
-        encode(img, enc)
+        encoder.new_image(img)
+        (enc, _) = encoder.compute()
```
Comment (breznak) [inline py/htm/examples/mnist.py]: WIP on MNIST, not yet tuned. I should revert these changes for now.

File: py/htm/encoders/eye.py [hunk 16]
```diff
@@ -0,0 +1,627 @@
...
+        """
+        self.output_diameter   = output_diameter
+        # Argument resolution_factor is used to expand the sensor array so that
+        # the fovea has adequate resolution.  After log-polar transform image
+        # is reduced by this factor back to the output_diameter.
+        self.resolution_factor = 3
+        self.retina_diameter   = int(self.resolution_factor * output_diameter)
+        # Argument fovea_scale  ... proportion of the image (ROI) which will be covered (seen) by
+        # high-res fovea (parvo pathway)
+        self.fovea_scale       = 0.177
```
Comment (breznak) [inline py/htm/encoders/eye.py]: I have previously misinterpreted this and self.scale. Rename and change this to `fovea_diameter` to be clearer?

File: py/htm/encoders/eye.py [hunk 17]
```diff
@@ -0,0 +1,627 @@
...
+        if color is True:
+          assert(sparsityParvo > 0)
+        self.color = color
+
+
+        self.retina = cv2.bioinspired.Retina_create(
+            inputSize            = (self.retina_diameter, self.retina_diameter),
+            colorMode            = color,
+            colorSamplingMethod  = cv2.bioinspired.RETINA_COLOR_BAYER,
+            useRetinaLogSampling = True,)
```
Comment (breznak) [inline py/htm/encoders/eye.py]: @ctrl-z-9000-times please compare with this on/off. Can it replace our manual log-polar transformation?

File: py/htm/encoders/eye.py [hunk 18]
```diff
@@ -0,0 +1,627 @@
...
+        else:
+            y_offset = 0
+        x_shape, y_shape, color_depth = image_slice.shape
+        roi[x_offset:x_offset+x_shape, y_offset:y_offset+y_shape] = image_slice
+
+        # Rescale the ROI to remove the scaling effect.
+        roi.resize( (self.retina_diameter, self.retina_diameter, 3))
+
+        # Mask out areas the eye can't see by drawing a circle boarder.
+        # this represents the "shape" of the sensor/eye (comment out to leave rectangural)
```
Comment (breznak) [inline py/htm/encoders/eye.py]: ok to crop to circular region here (and not only in the visualization)? Makes encoder see only ROI as the inner circle.

File: py/htm/encoders/eye.py [hunk 19]
```diff
@@ -0,0 +1,627 @@
...
+        if position is not None:
+          self.position = position
+        if rotation is not None:
+          self.orientation=rotation
+        if scale is not None:
+          self.scale=scale
+
+        # apply field of view (FOV), rotation
+        self.roi = self._crop_roi()
+        self.roi = self.rotate_(self.roi, self.orientation)
```
Comment (breznak) [inline py/htm/encoders/eye.py]: apply rotation to the image itself, instead of separately to output, visualizations, etc

File: py/htm/encoders/eye.py [hunk 20]
```diff
@@ -0,0 +1,627 @@
...
+    def small_random_movement(self):
+        """returns small difference in position, rotation, scale.
+           This is naive "saccadic" movements.
+        """
+        max_change_angle = (2*3.14159) / 500
+        self.position = (
+            self.position[0] + random.gauss(1, .75),
+            self.position[1] + random.gauss(1, .75),)
+        self.orientation += random.uniform(-max_change_angle, max_change_angle)
+        self.scale = 1
```
Comment (breznak) [inline py/htm/encoders/eye.py]: TODO randomize the scale too

File: py/htm/encoders/eye.py [hunk 21]
```diff
@@ -0,0 +1,627 @@
...
+        for img_path in images:
+            eye.reset()
+            print("Loading image %s"%img_path)
+            eye.new_image(img_path)
+            eye.fovea_scale = 0.177 #TODO find which value?
+            #eye.center_view()
+            eye.position=(400,400)
+            for i in range(10):
+                pos,rot,sc = eye.small_random_movement()
+                sc=1.0 #FIXME broken (only plot?) whenever scale != 1.0
```
Comment (breznak) [inline py/htm/encoders/eye.py]: FIXME changing scale breaks things (plot?)

Comment (breznak) [reply py/htm/encoders/eye.py]: fixed, caused by numerical computation, use cv2.resize instead of np.resize()
